### PR TITLE
cpu/sam0: unified rtt configuration

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

This PR is split from https://github.com/RIOT-OS/RIOT/pull/13874. It unifies `rtt` configuration for `sam0` and make the `RTT_FREQYENCY` configurable.

As in #13874 max and min values for the RTT are documented.

I've notices that `RT&DEV` are unused here, but I would rather remove in a follow up. I did remove `RTT_IRQ_PRIORITY` since it was unused and misleading.

### Testing procedure

Test for one board of every subfamily

- [x] samd21
- [x] saml1x
- [x] saml21
- [x] samd5x

i don't have any board of the other 3 ATM, maybe @benpicco @dylad can help?

- `tests/periph_rtt` should still pass on `sam0` boards with different frequencies

`BOARD=samr21-xpro make -C tests/periph_rtt flash test --no-print-directory -j3`
```
2020-06-17 12:50:59,314 # START
2020-06-17 12:50:59,321 # main(): This is RIOT! (Version: 2020.07-devel-1252-g69761-dev/sam0_unfied_rtt)
2020-06-17 12:50:59,321 #
2020-06-17 12:50:59,323 # RIOT RTT low-level driver test
2020-06-17 12:50:59,328 # This test will display 'Hello' every 5 seconds
2020-06-17 12:50:59,328 #
2020-06-17 12:50:59,330 # Initializing the RTT driver
2020-06-17 12:50:59,332 # RTT now: 2
2020-06-17 12:50:59,335 # Setting initial alarm to now + 5 s (163842)
2020-06-17 12:50:59,340 # Done setting up the RTT, wait for many Hellos
2020-06-17 12:51:04,331 # Hello
2020-06-17 12:51:09,331 # Hello
2020-06-17 12:51:14,331 # Hello
2020-06-17 12:51:19,331 # Hello
2020-06-17 12:51:24,331 # Hello
2020-06-17 12:51:29,331 # Hello
2020-06-17 12:51:34,331 # Hello
2020-06-17 12:51:39,331 # Hello

```

`CFLAGS+=-DRTT_FREQUENCY=1024 BOARD=samr21-xpro make -C tests/periph_rtt clean flash test --no-print-directory -j3`

```
2020-06-17 12:51:46,588 # START
2020-06-17 12:51:46,594 # main(): This is RIOT! (Version: 2020.07-devel-1252-g69761-dev/sam0_unfied_rtt)
2020-06-17 12:51:46,595 #
2020-06-17 12:51:46,597 # RIOT RTT low-level driver test
2020-06-17 12:51:46,601 # This test will display 'Hello' every 5 seconds
2020-06-17 12:51:46,602 #
2020-06-17 12:51:46,604 # Initializing the RTT driver
2020-06-17 12:51:46,605 # RTT now: 0
2020-06-17 12:51:46,609 # Setting initial alarm to now + 5 s (5120)
2020-06-17 12:51:46,613 # Done setting up the RTT, wait for many Hellos
2020-06-17 12:51:51,606 # Hello
2020-06-17 12:51:56,606 # Hello
2020-06-17 12:52:01,606 # Hello
2020-06-17 12:52:06,606 # Hello
2020-06-17 12:52:11,606 # Hello
```

### CODE REVIEW

File: cpu/saml21/include/periph_cpu.h [hunk 1]
```diff
@@ -73,6 +73,20 @@ typedef enum {
  */
 #define DAC_NUMOF           (2)
 
+/**
+ * @name    Real time counter configuration
+ * @{
+ */
+#define RTT_DEV             RTC->MODE0
+#define RTT_IRQ             RTC_IRQn
+#define RTT_ISR             isr_rtc
```
Comment (benpicco) [inline cpu/saml21/include/periph_cpu.h:82]: those are unused

File: cpu/saml21/include/periph_cpu.h [hunk 2]
```diff
@@ -73,6 +73,20 @@ typedef enum {
...
+
+#define RTT_MAX_VALUE       (0xffffffff)
```
Comment (benpicco) [inline cpu/saml21/include/periph_cpu.h:84]: That could be moved to `cpu/sam0_common/include/periph_cpu_common.h`

File: cpu/saml21/include/periph_cpu.h (hunk 1 above)
Comment (fjmolinas) [reply cpu/saml21/include/periph_cpu.h:82]: `RTT_IRQ` is used, I can remove the other two if you want. As well as `RTC_DEV`

File: cpu/saml21/include/periph_cpu.h (hunk 2 above)
Comment (fjmolinas) [reply cpu/saml21/include/periph_cpu.h:84]: Does it make splitting of all the definitions that much? I would rather look at everything in one header.

File: cpu/saml21/include/periph_cpu.h (hunk 1 above)
Comment (benpicco) [reply cpu/saml21/include/periph_cpu.h:82]: I can't find a use of `RTT_IRQ` inside `cpu/sam0_common`

File: cpu/saml21/include/periph_cpu.h (hunk 2 above)
Comment (benpicco) [reply cpu/saml21/include/periph_cpu.h:84]: Unlike `RTT_FREQUENCY`, this is not configurable but fixed for all sam0 CPUs.

File: cpu/saml21/include/periph_cpu.h (hunk 1 above)
Comment (fjmolinas) [reply cpu/saml21/include/periph_cpu.h:82]: Master was using RTC_IRQ https://github.com/RIOT-OS/RIOT/blob/beaec4c2ed906339e4cf3ea8b694bad6c920bb0b/cpu/sam0_common/periph/rtt.c#L127 I changed it to RTT_IRQ, so you would only find the use in this PR.

Comment (benpicco) [reply cpu/saml21/include/periph_cpu.h:82]: But `RTC_IRQn` is the correct interrupt from `IRQn_Type`.

Comment (fjmolinas) [reply cpu/saml21/include/periph_cpu.h:82]: Ok. I'll remove.

File: cpu/saml21/include/periph_cpu.h (hunk 2 above)
Comment (fjmolinas) [reply cpu/saml21/include/periph_cpu.h:84]: Yes, but that is not the argument I'm using, I don't want to be looking for `RTT` related  values over multiple headers. I don't think we gain anything by splitting it out.

**Comment by fjmolinas** [comment]
I'll trigger a build to see there is some conflict and the fix the changes requested

**Comment by benpicco** [comment]
Murdock is reasonably happy.

File: cpu/sam0_common/periph/rtt.c [hunk 3]
```diff
@@ -36,18 +36,28 @@
...
 #endif
 
+#ifdef REG_RTC_MODE0_CTRLA
+#define RTC_MODE0_PRESCALER       \
+    (__builtin_ctz(RTT_CLOCK_FREQUENCY / RTT_FREQUENCY) << \
+    RTC_MODE0_CTRLA_PRESCALER_Pos)
+#else
+#define RTC_MODE0_PRESCALER       \
+    (__builtin_ctz(RTT_CLOCK_FREQUENCY / RTT_FREQUENCY) << \
+    RTC_MODE0_CTRL_PRESCALER_Pos)
```
Comment (fjmolinas) [inline cpu/sam0_common/periph/rtt.c:46]: Tested on iotlab for `samr30-xpro` and figure out there is a difference  between `samd21` and `saml21` I'll add a patch.

Comment (fjmolinas) [reply cpu/sam0_common/periph/rtt.c:46]: image

image

Comment (fjmolinas) [reply cpu/sam0_common/periph/rtt.c:46]: Sam applies for the other sub-families.

**Comment by fjmolinas** [comment]
Tested on `sam30-xpro` on iotlab:

<details><summary>IOTLAB_NODE=samr30-2.saclay.iot-lab.info BOARD=samr30-xpro make -C tests/periph_rtt flash test --no-print-directory -j3</summary>

```
main(): This is RIOT! (Version: 2020.07-devel-1254-gbeaec-pr_sam0_unified_rtt)

RIOT RTT low-level driver test
This test will display 'Hello' every 5 seconds

Initializing the RTT driver
RTT now: 0
Setting initial alarm to now + 5 s (163840)
Done setting up the RTT, wait for many Hellos
Hello
Hello
Hello
Hello
Hello

```
</details>

<details><summary>CFLAGS+=-DRTT_FREQUENCY=1024 IOTLAB_NODE=samr30-3.saclay.iot-lab.info BOARD=samr30-xpro make -C tests/periph_rtt flash test
make: Entering directory '/home/francisco/workspace/RIOT/tests/periph_rtt'
</summary>

```
READY
s
START
main(): This is RIOT! (Version: 2020.07-devel-1254-gbeaec-pr_sam0_unified_rtt)

RIOT RTT low-level driver test
This test will display 'Hello' every 5 seconds

Initializing the RTT driver
RTT now: 0
Setting initial alarm to now + 5 s (5120)
Done setting up the RTT, wait for many Hellos
Hello
Hello
Hello
Hello
Hello
```
</details>

**Comment by fjmolinas** [comment]
@benpicco do you think you could test on `samd5x` and `saml1x` if you have any?

File: cpu/saml21/include/periph_cpu.h (hunk 1 above)
Comment (fjmolinas) [reply cpu/saml21/include/periph_cpu.h]: Removed!

File: cpu/sam0_common/periph/rtt.c (hunk 3 above)
Comment (fjmolinas) [reply cpu/sam0_common/periph/rtt.c]: c11dcf7 addresses this

**Comment by benpicco** [comment]
Works as expected

<details><summary>same54-xpro</summary>

```
2020-06-18 23:57:44,498 # RIOT RTT low-level driver test
2020-06-18 23:57:44,502 # This test will display 'Hello' every 5 seconds
2020-06-18 23:57:44,502 # 
2020-06-18 23:57:44,504 # Initializing the RTT driver
2020-06-18 23:57:44,505 # RTT now: 0
2020-06-18 23:57:44,509 # Setting initial alarm to now + 5 s (163840)
2020-06-18 23:57:44,513 # Done setting up the RTT, wait for many Hellos
2020-06-18 23:57:49,505 # Hello
2020-06-18 23:57:54,505 # Hello
2020-06-18 23:57:59,506 # Hello
2020-06-18 23:58:04,506 # Hello
```
</details>

<details><summary>saml10-xpro</summary>

```
2020-06-19 00:01:55,678 # RIOT RTT low-level driver test
2020-06-19 00:01:55,682 # This test will display 'Hello' every 5 seconds
2020-06-19 00:01:55,683 # 
2020-06-19 00:01:55,685 # Initializing the RTT driver
2020-06-19 00:01:55,686 # RTT now: 0
2020-06-19 00:01:55,690 # Setting initial alarm to now + 5 s (163840)
2020-06-19 00:01:55,694 # Done setting up the RTT, wait for many Hellos
2020-06-19 00:02:00,686 # Hello
2020-06-19 00:02:05,686 # Hello
2020-06-19 00:02:10,686 # Hello
```
</details>

**Comment by fjmolinas** [comment]
@benpicco all green here!

**Comment by fjmolinas** [comment]
Thanks for reviewing and testing @benpicco!

**Pull request merged**